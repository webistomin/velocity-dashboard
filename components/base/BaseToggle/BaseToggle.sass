+b('base-toggle')
  width: rem(50px)
  height: rem(50px)
  position: relative

  &:hover .base-toggle__line
    background-color: var(--color-primary)

    &:last-of-type
      width: rem(18px)

  +m('active')
    .base-toggle__line
      &:first-of-type
        transform: translateY(rem(6px)) rotate(45deg)

      &:nth-of-type(2)
        transform: translateX(-50%)
        opacity: 0

      &:last-of-type
        width: rem(18px)
        transform: translateY(rem(-6px)) rotate(-45deg)

  +e('line')
    display: block
    position: absolute
    width: rem(18px)
    height: rem(2px)
    background-color: var(--color-gray-4)
    left: rem(16px)
    transition: background-color 0.2s ease-out, width 0.2s ease-out, transform 0.2s ease-out, opacity 0.2s ease-out

    &:first-of-type
      top: calc(50% - 0.6rem)

    &:nth-of-type(2)
      top: 50%

    &:last-of-type
      width: rem(11px)
      top: calc(50% + 0.6rem)
