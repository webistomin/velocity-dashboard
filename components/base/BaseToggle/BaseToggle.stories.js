import { text, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import BaseToggle from './BaseToggle';

export default {
  title: 'BaseToggle',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
    },
  },
};

export const Default = () => ({
  components: { BaseToggle },
  props: {
    size: {
      type: String,
      default: text('size', 'xs'),
    },
    isActive: {
      type: Boolean,
      default: boolean('isActive', false),
    },
  },
  methods: {
    onClick: action('emit click'),
  },
  template: `<div style="display: flex; padding: 40px;">
      <base-toggle
        @click="onClick()"
        :size="this.size"
        :isActive="this.isActive"
      />
  </div>`,
});
