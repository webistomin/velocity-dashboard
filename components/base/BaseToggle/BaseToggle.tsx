import { VueComponent } from 'types/vue-components';
import { Component, Emit, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import './BaseToggle.sass';

export type IBaseToggleSize = 'xs' | 's';

export interface IBaseToggleProps {
  size?: IBaseToggleSize;
  isActive: boolean;
  onClick?: () => void;
}

@Component({
  name: 'BaseToggle',
})
export default class BaseToggle extends VueComponent<IBaseToggleProps> {
  @Prop({ default: 'xs' })
  private readonly size!: IBaseToggleProps['size'];

  @Prop({ default: false, required: true })
  private readonly isActive!: IBaseToggleProps['isActive'];

  @Emit('click')
  public onToggleClick() {}

  public render(): VNode {
    const { size, isActive } = this;
    return (
      <button
        data-jest='base-toggle'
        class={`base-toggle btn base-toggle_size_${size} ${isActive ? 'base-toggle_active' : ''}`}
        type='button'
        aria-label='Toggle'
        onClick={this.onToggleClick}>
        <span class='base-toggle__wrapper'>
          <span class='base-toggle__line' />
          <span class='base-toggle__line' />
          <span class='base-toggle__line' />
        </span>
      </button>
    );
  }
}
