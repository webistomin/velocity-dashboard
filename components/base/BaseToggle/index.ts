import BaseToggle from './BaseToggle';

export default BaseToggle;
