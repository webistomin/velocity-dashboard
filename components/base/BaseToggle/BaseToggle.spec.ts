import { mount } from '@vue/test-utils';
import BaseToggle from './BaseToggle';

describe('BaseToggle', () => {
  const toggleSelector = "[data-jest='base-toggle']";

  const factory = (propsData = {}, options = {}) => {
    return mount(BaseToggle, {
      propsData: {
        isActive: false,
        ...propsData,
      },
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Emit "input" event', () => {
    const wrapper = factory();
    const btn = wrapper.find(toggleSelector);
    btn.trigger('click');
    expect(wrapper.emitted().click);
  });

  it('Set class from color prop', () => {
    const expectedResult = ['base-toggle_active'];
    const wrapper = factory({
      isActive: true,
    });
    const classes = wrapper.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });
});
