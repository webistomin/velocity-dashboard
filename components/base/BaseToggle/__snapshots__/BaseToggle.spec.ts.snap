// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseToggle Matches snapshot 1`] = `"<button data-jest=\\"base-toggle\\" type=\\"button\\" aria-label=\\"Toggle\\" class=\\"base-toggle btn base-toggle_size_xs \\"><span class=\\"base-toggle__wrapper\\"><span class=\\"base-toggle__line\\"></span><span class=\\"base-toggle__line\\"></span><span class=\\"base-toggle__line\\"></span></span></button>"`;
